frontend/device/android/device.lua

```diff
@@ -375,6 +374,13 @@ function Device:canExecuteScript(file)
     end
 end
 
+function Device:isValidPath(path)
+    if path:find("^/sdcard") ~= nil then
```

Comment:
This doesn't belong here.

Reply:
`/sdcard` is a symbolic link which points to ` /storage/sdcard1` on Android. So it the path begins with /sdcard it points to our sandbox.

Reply:
Or do you mean it should be in `android.isPathInsideSandbox`

Reply:
> Or do you mean it should be in android.isPathInsideSandbox

I mean that's where it should be if it weren't, but that's its very raison d'être! ;-)  All it does is check if the path starts with `Environment.getExternalStorageDirectory()`, which will return a value like `/sdcard`, `/storage/sdcard0`, `/storage/sdcard1`, etc. as appropriate.

Reply:
> which will return a value like /sdcard, /storage/sdcard0, /storage/sdcard1, etc. as appropriate.

`Environment.getExternalStorageDirectory()`  will return the real path (in that case /storage/sdcard1). `/sdcard` is just a bogus path that points there on 95% of devices 90% of the time :p 

So, yeah. If a path starts with `/sdcard` the thing that will work on 100% of devices would be get the real path of the symbolic link and check if that matches the path returned by `Environment.getExternalStorageDirectory()`. I agree that can be moved to `android.isPathInsideSandbox`

Reply:
In any case, either it's useful and it should be in luajit-launcher or it shouldn't be at all.

Reply:
Is there really a use case to *hardcode* /sdcard?

Reply:
where? :P 

All started in https://github.com/koreader/koreader/pull/6813#issuecomment-716124026 (and following 2 comments). Then @zwim did wrote the code and I answered in https://github.com/koreader/koreader/pull/6813#discussion_r513034218

The main usage *here* is to validate a input dialog entry that a user typed. Users don't know the real path and they're used to know that koreader and other apps have "folders" in /sdcard.

Anyways, the initial implementation (check if the path is a symbolic link using lfs) cannot be done in luajit-launcher, as we're not having access to lua modules there. As @zwim talked about how his proposal would work only sometimes, I suggested current implementation an asume that if a path starts with /sdcard this path will be a symlink to a path that it's under external storage path.

So 4 ways to solve that:

- do not validate user input at all
- validate user input properly with:
    1. check if path starts with `/sdcard`
    2. if that happens resolve the path using lfs
    3. check the resolved path against insideSandbox
- leave the stuff as is (without resolving the real path, just assuming that /sdcard indeed points to external storage). In that case we can move the function to luajit-launcher
- Implement the code to check if /sdcard == storage path in kotlin and update things accordly.

Reply:
As always, this is full of android sh*t

Reply:
Er, I'm talking about *this line*, which is fully separate from any other comments. ;-)

It's a matter of maintainability. Stick this [here](https://github.com/koreader/android-luajit-launcher/blob/f2d946b3b49e728272df4cb56185a2fe57cdb4ff/app/src/org/koreader/launcher/MainActivity.kt#L486-L490) or it will be super confusing to track down how the program actually works. There's no lfs here or some such.

Whether you stick it in plain library-less Lua in this very form or in Java/Kotlin with startsWith/regex/whatever all works the same

Reply:
In that case the following should work:

```diff
diff --git a/app/src/org/koreader/launcher/MainActivity.kt b/app/src/org/koreader/launcher/MainActivity.kt
index 7c0039c..c4adc95 100644
--- a/app/src/org/koreader/launcher/MainActivity.kt
+++ b/app/src/org/koreader/launcher/MainActivity.kt
@@ -21,7 +21,7 @@ import androidx.core.app.ActivityCompat
 import androidx.core.content.ContextCompat
 import org.koreader.launcher.interfaces.JNILuaInterface
 import org.koreader.launcher.utils.*
-import java.io.IOException
+import java.io.*
 import java.util.*
 
 class MainActivity : NativeActivity(), JNILuaInterface,
@@ -483,9 +483,15 @@ class MainActivity : NativeActivity(), JNILuaInterface,
         }
     }
 
+    @SuppressLint("SdCardPath")
     override fun isPathInsideSandbox(path: String?): Int {
         return path?.let {
-            if (it.startsWith(device.externalStorage)) 1 else 0
+            val realPath = if (it.startsWith("/sdcard")) {
+                it.replace("/sdcard", File("/sdcard").canonicalPath)
+            } else {
+                it
+            }
+            if (realPath.startsWith(device.externalStorage)) 1 else 0
         } ?: 0
     }
```

@zwim: could you test it?

Reply:
@pazos: I am afraid, it does not work. I have deleted `path:find("^/sdcard")` thing in `android/device.lua`.

`File("/sdcard").canonicalPath` points to `/mnt/media_rw/sdcard1` and 
`device.externalStorage` is `/storage/sdcard1` on my device.

When I replace `it.replace("/sdcard", File("/sdcard").canonicalPath)` with `it.replace("sdcard",device.externalStorage)` it works.

Reply:
what a shame, based on File documentation I would expect them to match.

Reply:
Maybe this is due to Tolino Firmware 14.0.1 (which is some kind of Android 4.4.2).
Here I get with `ls -al /sd*`
```
lrwxrwxrwx root     root              2020-11-05 11:40 sdcard -> /storage/sdcard1
```
and with `ls -al /storage/*`
```
lrwxrwxrwx root     root              2020-11-05 11:40 extsd -> /mnt/media_rw/extsd
lrwxrwxrwx root     root              2020-11-05 11:40 sdcard1 -> /mnt/media_rw/sdcard1
lrwxrwxrwx root     root              2020-11-05 11:40 udisk -> /mnt/media_rw/udisk
```